src/libs/sfdk/virtualmachine.cpp

```diff
@@ -0,0 +1,798 @@
+{
+    return d_func()->connection->isLockedDown();
+}
+
+int VirtualMachine::memorySizeMb() const
+{
+    Q_D(const VirtualMachine);
+    QTC_ASSERT(d->initialized(), return false);
+    return d->virtualMachineInfo.memorySizeMb;
+}
+
+void VirtualMachine::setMemorySizeMb(int memorySizeMb, const QObject *context,
+        const Functor<bool> &functor)
+{
+    Q_D(VirtualMachine);
+    QTC_CHECK(isLockedDown());
+
+    const QPointer<const QObject> context_{context};
+    d->doSetMemorySizeMb(memorySizeMb, this, [=](bool ok) {
+        if (ok && d->virtualMachineInfo.memorySizeMb != memorySizeMb) {
+            d->virtualMachineInfo.memorySizeMb = memorySizeMb;
+            VirtualMachineInfoCache::insert(uri(), d->virtualMachineInfo);
+            emit memorySizeMbChanged(memorySizeMb);
+        }
+        if (context_)
+            functor(ok);
+    });
+}
+
+int VirtualMachine::availableMemorySizeMb()
+{
+#ifdef Q_OS_WIN
+    MEMORYSTATUSEX status;
+    status.dwLength = sizeof(status);
+    GlobalMemoryStatusEx(&status);
+    return status.ullTotalPhys >> 20;
+#else
+    long pages = sysconf(_SC_PHYS_PAGES);
+    long page_size = sysconf(_SC_PAGE_SIZE);
+    return (pages * page_size) >> 20;
+#endif
+}
+
+int VirtualMachine::cpuCount() const
+{
+    Q_D(const VirtualMachine);
+    QTC_ASSERT(d->initialized(), return false);
+    return d->virtualMachineInfo.cpuCount;
+}
+
+void VirtualMachine::setCpuCount(int cpuCount, const QObject *context, const Functor<bool> &functor)
+{
+    Q_D(VirtualMachine);
+    QTC_CHECK(isLockedDown());
+
+    const QPointer<const QObject> context_{context};
+    d->doSetCpuCount(cpuCount, this, [=](bool ok) {
+        if (ok && d->virtualMachineInfo.cpuCount != cpuCount) {
+            d->virtualMachineInfo.cpuCount = cpuCount;
+            VirtualMachineInfoCache::insert(uri(), d->virtualMachineInfo);
+            emit cpuCountChanged(cpuCount);
+        }
+        if (context_)
+            functor(ok);
+    });
+}
+
+int VirtualMachine::availableCpuCount()
+{
+    return QThread::idealThreadCount();
+}
+
+int VirtualMachine::vdiCapacityMb() const
+{
+    Q_D(const VirtualMachine);
+    QTC_ASSERT(d->initialized(), return false);
+    return d->virtualMachineInfo.vdiCapacityMb;
+}
+
+void VirtualMachine::setVdiCapacityMb(int vdiCapacityMb, const QObject *context,
+        const Functor<bool> &functor)
+{
+    Q_D(VirtualMachine);
+    QTC_CHECK(isLockedDown());
+
+    const QPointer<const QObject> context_{context};
+    d->doSetVdiCapacityMb(vdiCapacityMb, this, [=](bool ok) {
+        if (ok && d->virtualMachineInfo.vdiCapacityMb != vdiCapacityMb) {
+            d->virtualMachineInfo.vdiCapacityMb = vdiCapacityMb;
+            VirtualMachineInfoCache::insert(uri(), d->virtualMachineInfo);
+            emit vdiCapacityMbChanged(vdiCapacityMb);
+        }
+        if (context_)
+            functor(ok);
+    });
+}
+
+bool VirtualMachine::hasPortForwarding(quint16 hostPort, QString *ruleName) const
```

Comment:
I'm not sure the name of this method is the best possible. I get the feeling that the main purpose of the method is to get the name of the rule instead of just checking the existence of the rule.

Reply:
The main purpose is really to check if a port is used or free to use. The optional out-argument ruleName exists to allow verification (in the former case) that the rule is the expected one.

Reply:
Now I just noticed that the method is not even called anywhere. Is it even needed at all?

Reply:
It is used by the sensors emulation plugin developed by FRUCT https://github.com/sailfishos/sailfish-qtcreator/pull/165/files#diff-af529b4089f9cb42afa0b6dca536bbf2R146

Reply:
Ah, ok then. I would have done it the other way around, by including the additional optional check in hasPortForwarding() instead of having the caller do it. But I guess that ship has already sailed :-/

Reply:
What you prefer would not allow to query the rule name :)

Reply:
> What you prefer would not allow to query the rule name :)

That was exactly why I started asking about the method name in the first place. If I want to query the rule name, "hasPortForwarding()" is not the first method I will look at. I would first try to find if there is a method called "portForwadingRuleName()" or something like that. But then you wrote that it's not actually used for that but instead for checking that the name is what expected, i.e. the name is already known. In which case the caller is not interested in querying the rule name but verifying that it really is what it should be. So which is it? Is it used for querying the rule name, or checking that the rule name matches? In the first case the method name is not the best possible, and in the latter case the implementation could be done here, instead of everywhere where this method is called.

Reply:
```
QString ruleName;
if (vm->hasPortForwarding(PORT, &ruleName)) {
     if (ruleName != RULE_NAME)
           throw new StolenPortError(ruleName);
} else {
   vm->addPortForwarding(RULE_NAME, ...);
}
```

Reply:
Nitpick: in the original implementation, it saved one VBoxManage invocation, hence the design :)